docs/enhancements/cancel-day-2-host-installation.md

```diff
@@ -0,0 +1,117 @@
+cancellation of installation for a day-2 host.
+
+## Motivation
+
+There would be times a user is installing a day-2 host and in the middle of installation, they need to
+change something, or the install becomes stuck because of an error. There are also instances where a
+user might need to provision and deprovision a host quickly as it may no longer be needed, such as when
+autoscaler is used. 
+
+When a host starts installation, there's no way to remove it from a cluster until the host installation
+times out and errors out or it finishes installation. These can take a very long time, and it can be a
+frustrating experiences for users.
+
+### Goals
+
+- Once a day-2 host starts installation, the installation can be cancelled
+- There's a straight-forward way to trigger cancellation
+- Host whose installations are cancelled can be reclaimed back to their InfraEnvs and be reused
+- Host cancellation should not affect other hosts in the cluster and should not affect the cluster itself
+- Host cancellation should happen within a reasonable amount of time in order to support dynamic scaling 
+
+### Non-Goals
+
+- Hosts that are installed in a day-1 cluster is out of scope of this enhancement
+- Cluster installation cancellation is out of scope
+- The way reclaim/unbind works should not be affected by these changes
+- This enhancement is focused solely on Assisted-Service offerings using Kubernetes API (SaaS should not be affected by this)
+
+## Proposal
+
+Allow cancelling an installing day-2 host through Assisted-Service's Kubernetes API.
+
+### User Stories
+
+#### Story 1
+
+As a cluster admin, when I create a hosted cluster using Hypershift, I'd like to bind and install worker hosts into my cluster. 
+However, if I want to remove the host from the cluster before they finish installing, I should be able to do so and reuse the host for another cluster.
+
+#### Story 2
+
+As a cluster admin using Hypershift, I want to use the autoscaler.
+When the scaling happens and hosts are bound/unbound from the cluster, the hosts should bind and unbind 
+easily from the cluster.
+
+#### Story 3
+
+As a cluster admin, if I install a day-2 host to a cluster, and I see the host installation fails from
+inside the host. I should be able to cancel the installation once it starts and quickly reset the host.
+
+### Implementation Details/Notes/Constraints [optional]
+
+There is an already existing cancel host installation API which maps to the unbind and reclaim stages.
+This is currently only being used when a day-1 cluster installation is cancelled. 
+
+We can use the existing host installation cancellation API, and allow the install cancellation 
+to happen from the Kubernetes API by removing the existing Agent webhook that blocks removing
+the cluster refrence while the Agent is installing.
+
+In the Agent controller, detect if the Agent becomes unbound from cluster while installing.
+If it is unbound, then allow it to be cancelled. 
```

Comment:
How will this work for hosts that are part of an infraenv that references a cluster (early-binding)?

Is that a case we want to handle? Maybe for those hosts deleting the agent could be the cancellation trigger (if that's not already happening)?

Reply:
Good point! So I see that there's nothing in the webhook preventing the Agent from being deleted

In the Agent controller, it just tries to remove the spoke resources ([src](https://github.com/openshift/assisted-service/blob/1dc66326d304c3c2bffe346de6e2ab278f0d0ac9/internal/controller/controllers/agent_controller.go#L501)) (which shouldn't be an issue)

And the Agent/host from the DB ([src](https://github.com/openshift/assisted-service/blob/master/internal/bminventory/inventory.go#L3459)) with nothing gating deletion.

I'm not sure cancellation will happen 🤔 (though I might've missed something) so this seems like another case to consider adding a cancellation when the Agent is being deleted, but is not fully installed.